### DISCUSSION TITLE
Understanding soldier promotion?

### DISCUSSION THREAD

**Discussion by juj** [nicymike/serflings#82] [General]
I've been trying to figure out how the soldier promotion mechanism works - naturally to try to get soldiers promoted to captains as quickly as possible.

So far what I've realized:
 - soldiers are always only promoted while they reside in the HQ(?), never while they are stationed in guard structures
 - from past memory, promoting a soldiers required (only?) gold bars (== coins) from the goldsmith
 - there is a button in the options menu to make strong soldiers rotate out and weak soldiers rotate back in to HQ for training

However I have run into a situation where there are several gold bars in the HQ, and unpromoted soldiers. Is there something that I am missing? Or is it just that promoting soldiers to captains takes a long time? (there is some kind of promotion interval timer?)

**Comment by nicymike** [comment]
First about your three items:
1. Knights are beeing promoted if they are inside the castle, any stock or any military building. I will talk more about chances later.
2. Promoting knights does not require gold, just time.
3. You are right about the button. It is for moving weak knights out of military buildings to be replaced by better ones (if available).

Approximately every 60 seconds, on default game speed, a random number will be generated to determine the chance for the promotion of a single knight. If the chance is high enough, the promotion will be done. This is done separately for every knight with its own random number.

The chances for the individual grades of a knight sitting inside a stock or the castle are as follows (65535 is the range for every random number in settlers 1):
1: 4000 / 65535 = 6.1%
2: 2000 / 65535 = 3.1%
3: 1000 / 65535 = 1.5%
4:  500 / 65535 = 0.76%
5: Final Grade, no promotion

So for stocks or the castle the chances are calculated by using the numbers 4000, 2000, 1000 and 500. Stocks and the castle provide the biggest chances for promotion.

The numbers for the military buildings are:
garrison: 2000, 1000, 500, 250 (half of stock/castle)
watch tower: 1000, 500, 250, 125 (half of garrison)
guard room: 250, 125, 62, 31 (quarter of watch tower)

As you can see it is possible for a knight to be promoted in military buildings. However the chances are heavily dropping for guard rooms. The knight exchange remains the preferred way to get higher grades near the border.

**Reply by juj** [reply]
Thanks, that explains the logic! So if one wants to promote as efficiently as possible, one will want to bump up the # of soldiers kept in the castle option setting (which defaults to 3) immediately at level start, and set the civilian<->soldier recruit slider all the way to the soldier end immediately at game start. That way there will be most soldiers in the castle waiting for their random chances to get a promotion.

This also hints a better understanding towards the other aspect that I've been puzzled about: how does soldier's strength form?

I presumed before that the soldier's strength was all about their promotion level, and promotion came with coins.. but now I understand there are two axes: promotion axis and motivation(?) axis. Soldiers get promoted according to the random chance logic you described, and they gain motivation when they are fed gold coins(?)

I suppose all new soldiers in the castle start up with zero motivation (under the ZZzz "lazy" icon in the attack screen), and they gain motivation when they are in a military building while gold is delivered into it?

Is motivation somehow temporary? If gold is delivered to a soldier, but then gold delivery halts, do the soldiers slowly revert back to being lazy?

I suppose there is then some kind of formula in the fights to turn the promotion and motivation ranks into a "hit % chance"?

**Reply by nicymike** [reply]
> Thanks, that explains the logic! So if one wants to promote as efficiently as possible, one will want to bump up the # of soldiers kept in the castle option setting (which defaults to 3) immediately at level start, and set the civilian<->soldier recruit slider all the way to the soldier end immediately at game start. That way there will be most soldiers in the castle waiting for their random chances to get a promotion.

Yes, but keep in mind that this will possibly lead to no unemployed settlers to do other jobs (especially carriers) and to no knights occupying military buildings if you have less than requested castle guards.

> This also hints a better understanding towards the other aspect that I've been puzzled about: how does soldier's strength form?
> 
> I presumed before that the soldier's strength was all about their promotion level, and promotion came with coins.. but now I understand there are two axes: promotion axis and motivation(?) axis. Soldiers get promoted according to the random chance logic you described, and they gain motivation when they are fed gold coins(?)
> 
> I suppose all new soldiers in the castle start up with zero motivation (under the ZZzz "lazy" icon in the attack screen), and they gain motivation when they are in a military building while gold is delivered into it?
> 
> Is motivation somehow temporary? If gold is delivered to a soldier, but then gold delivery halts, do the soldiers slowly revert back to being lazy?

You are right about the promotion. It is all about chance over time. But you are wrong about motivation.

Motivation is not connected to a single knight. It depends on the amount of gold in military buildings, stocks and the castle and every knight has the same motivation.

The motivation (first number as percentage), along with the number of calculated gold (second number), can be seen in the knight settings.

![image](https://user-images.githubusercontent.com/19415000/184158954-4ad4b329-90cf-44ed-83cc-05ec6d44b135.png)

As seen in the picture I have 148 gold which leads to a motivation of 217%.

The ZZzz-icon on the attack screen has in fact nothing to do with motivation. 🙂 The icons on the attack screen only indicate the distance.

![image](https://user-images.githubusercontent.com/19415000/184161146-58ae650d-5a12-4731-84fd-0218896d641f.png)

> I suppose there is then some kind of formula in the fights to turn the promotion and motivation ranks into a "hit % chance"?

Yes, calculated as shown below.

Motivation for a knight on your own territory is always 100% regardless of attacking or defending.
Knight_grade is a number from 0 (first grade) to 4 (fifth grade).

Numberic values for motivation (if not on own territory) as seen in the knight settings:
0% = 1
25% = 1024
50% = 2048
75% = 3072
100% = 4096
200% = 8192
...
1600% = 65535

Calculated attackerScore: ((1024 << knight_grade) * Motivation) >> 16
Calculated defenderScore: (((((1024 << knight_grade) * Motivation) >> 16) + attackerScore) * Random(0-65535)) >> 16

if defenderScore < attackerScore
= attacker wins
else
= defender wins

Also important: If your motivation is above 100% it only counts if you don't fight on your own territory. This means your knights are weaker while defending your own buildings than attacking enemy buildings.
Also the calculation is done for every single fight. There is no health like in Settlers 2.

**Reply by dvilker** [reply]
Thanks for the explanation. I've been interested in this question since childhood!)
It turns out that using Save/Load/Load/.../ you can capture any object. Sooner or later you'll get lucky with the random.

**Reply by nicymike** [reply]
Yes, because the random numbers are rotated on every load to add more variety to the gameplay. In Serflings this is added on purpose while in the original game it has technical reasons.

**Comment by rtrzebinski** [comment]
> Motivation is not connected to a single knight. It depends on the amount of gold in military buildings, stocks and the castle and every knight has the same motivation.

@nicymike Thank you for these explanations, it's amazing to read these internal details after decades of playing a game.

Question I have regarding the quite - so does it matter whether the gold is delivered to military buildings at all? What does it change? The impact of gold in the castle is the same as the gold in military buildings?

**Reply by nicymike** [reply]
You are right, the impact is the same. It does not matter if gold is inside a military building, a stock or the castle. Only gold outside of a building is not counting.

I don't know why the game is designed to deliver gold to military buildings. Maybe it's because gold is safer inside military buildings than inside stocks. A stock cannot be defended after all.
